FV3/atmos_cubed_sphere/tools/coarse_grained_diagnostics.F90

```diff
@@ -1361,4 +1414,72 @@ subroutine compute_grid_coarse(is, ie, js, je, is_coarse, ie_coarse, &
     factor = Atm(tile_count)%coarse_graining%factor
     grid_coarse = Atm(tile_count)%gridstruct%grid(is:ie+1:factor,js:je+1:factor,:)
   end subroutine compute_grid_coarse
+
+  logical function is_derived_2d_field(coarse_diag)
+    type(coarse_diag_type), intent(in) :: coarse_diag
+
+    is_derived_2d_field = &
+           coarse_diag%pressure_level .gt. 0 &
+      .or. coarse_diag%vertically_integrated &
+      .or. coarse_diag%scaled_by_specific_heat_and_vertically_integrated &
+      .or. trim(coarse_diag%special_case) .ne. ''
+  end function is_derived_2d_field
+
+  subroutine get_hydrometeor_indexes(sphum, liq_wat, ice_wat, rainwat, snowwat, graupel)
+    integer, intent(out) :: sphum, liq_wat, ice_wat, rainwat, snowwat, graupel
+
+    sphum   = get_tracer_index (MODEL_ATMOS, 'sphum')
+    liq_wat = get_tracer_index (MODEL_ATMOS, 'liq_wat')
+    ice_wat = get_tracer_index (MODEL_ATMOS, 'ice_wat')
+    rainwat = get_tracer_index (MODEL_ATMOS, 'rainwat')
+    snowwat = get_tracer_index (MODEL_ATMOS, 'snowwat')
+    graupel = get_tracer_index (MODEL_ATMOS, 'graupel')
+  end subroutine get_hydrometeor_indexes
+
+  logical function is_multi_species_water_path(special_case)
+    character(len=64), intent(in) :: special_case
+
+    is_multi_species_water_path = &
+           trim(special_case) .eq. 'total_water_path' &
+      .or. trim(special_case) .eq. 'total_ice_water_path' &
+      .or. trim(special_case) .eq. 'total_liquid_water_path'
+  end function is_multi_species_water_path
+
+  subroutine compute_multi_species_water_path(is, ie, js, je, npz, n_prognostic, n_water, q, delp, case, result)
+    integer, intent(in) :: is, ie, js, je, npz, n_prognostic, n_water
+    real, intent(in) :: q(is:ie,js:je,1:npz,1:n_prognostic)
+    real, intent(in) :: delp(is:ie,js:je,1:npz)
+    character(len=64), intent(in) :: case
+    real, intent(out) :: result(is:ie,js:je)
+
+    real, allocatable :: total(:,:,:)
+    integer :: sphum, liq_wat, ice_wat, rainwat, snowwat, graupel
+
+    allocate(total(is:ie,js:je,1:npz))
+    total = 0.0
+
+    call get_hydrometeor_indexes(sphum, liq_wat, ice_wat, rainwat, snowwat, graupel)
+
+    if (trim(case) .eq. 'total_water_path') then
+      total = sum(q(is:ie,js:je,1:npz,1:n_water), dim=4)
+    elseif (trim(case) .eq. 'total_ice_water_path') then
+      if (ice_wat .gt. 0) then
```

Comment:
This would clarify my confusion earlier.
```suggestion
      ! if the indices are less than or equal to zero, that water type is unused
      if (ice_wat .gt. 0) then
```

Reply:
Indeed, good to add a comment here.  All set.